File: client/lib/arena/pickup.dart
import 'package:batufo/engine/tile_position.dart';
import 'package:batufo/rpc/generated/message_bus.pb.dart'
    show PackedPickup, PackedPickupType;

enum PickupType {
  Medkit,
  Shield,
  Bomb,
}

PickupType unpackPickupType(PackedPickupType data) {
  return PickupType.values[data.value];
}

PackedPickupType packPickupType(PickupType type) {
  return PackedPickupType.values[type.index];
}

class Pickup {
  final PickupType type;
  final TilePosition tilePosition;

  String get id => '$type:[${tilePosition.col},${tilePosition.row}]';

  Pickup(this.type, this.tilePosition);

  PackedPickup pack() {
    final packedType = packPickupType(type);
    final packedTilePosition = tilePosition.pack();

    return PackedPickup()
      ..type = packedType
      ..tilePosition = packedTilePosition;
  }

  static Pickup unpack(PackedPickup data) {
    final type = unpackPickupType(data.type);
    final tilePosition = TilePosition.unpack(data.tilePosition);

    return Pickup(type, tilePosition);
  }
}

File: backend/packages/core/src/arena/pickup.ts
import { strict as assert } from 'assert'
import { TilePosition } from './tile-position'
import {
  PackedPickup,
  PackedPickupType,
  PackedPickupTypeMap,
} from '../generated/message_bus_pb'

export enum PickupType {
  Medkit,
  Shield,
  Bomb,
}

function unpackPickupType(
  data: typeof PackedPickupType[keyof PackedPickupTypeMap]
): PickupType {
  return data
}

function packPickupType(
  type: PickupType
): typeof PackedPickupType[keyof PackedPickupTypeMap] {
  return type
}

export class Pickup {
  constructor(readonly type: PickupType, readonly tilePosition: TilePosition) {}

  pack(): PackedPickup {
    const packedType = packPickupType(this.type)
    const packedTilePosition = this.tilePosition.pack()

    const packedPickup = new PackedPickup()
    packedPickup.setType(packedType)
    packedPickup.setTileposition(packedTilePosition)
    return packedPickup
  }

  static unpack(data: PackedPickup): Pickup {
    const type = unpackPickupType(data.getType())

    const packedTilePosition = data.getTileposition()
    assert(packedTilePosition != null, 'pickup cannot be missing tilePosition')
    const tilePosition = TilePosition.unpack(packedTilePosition)

    return new Pickup(type, tilePosition)
  }
}
